pkg/opts/container.go

```diff
@@ -21,23 +21,49 @@ import (
 
 	"github.com/containerd/containerd"
 	"github.com/containerd/containerd/containers"
+	"github.com/containerd/containerd/errdefs"
 	"github.com/docker/docker/pkg/chrootarchive"
 	"github.com/docker/docker/pkg/system"
+	"github.com/opencontainers/image-spec/identity"
 	"github.com/pkg/errors"
 	"golang.org/x/sys/unix"
 )
 
+// WithImageUnpack guarantees that the image used by the container is unpacked.
+func WithImageUnpack(i containerd.Image) containerd.NewContainerOpts {
+	return func(ctx context.Context, client *containerd.Client, c *containers.Container) error {
+		if c.Snapshotter == "" {
+			return errors.New("no snapshotter set for container")
+		}
+		snapshotter := client.SnapshotService(c.Snapshotter)
+		diffIDs, err := i.RootFS(ctx)
+		if err != nil {
+			return errors.Wrap(err, "get image diff IDs")
+		}
+		chainID := identity.ChainID(diffIDs)
+		_, err = snapshotter.Stat(ctx, chainID.String())
+		if err == nil {
+			return nil
+		}
+		if !errdefs.IsNotFound(err) {
+			return errors.Wrap(err, "stat snapshot")
+		}
+		// Unpack the snapshot.
+		return i.Unpack(ctx, c.Snapshotter)
```

Comment:
Maybe wrap the unpack error here with "unpack snapshot"?

Reply:
Will do.

Reply:
Done